File: LC012IntegerToRoman.java
package leetcode;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/integer-to-roman/
 */
public class LC012IntegerToRoman {
    public class Solution {
        private String[] digits = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};
        private int[] weights = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};

        public String intToRoman(int num) {
            StringBuilder roman = new StringBuilder();
            for (int i = 0; i < digits.length; i++) {
                int multiple = num / weights[i];
                for (int j = 0; j < multiple; j++) {
                    roman.append(digits[i]);
                }
                num -= multiple * weights[i];
            }
            return roman.toString();
        }
    }

    @Test
    public void test_1234() throws Exception {
        assertEquals("MCCXXXIV", new Solution().intToRoman(1234));
    }

    @Test
    public void test_9() throws Exception {
        assertEquals("IX", new Solution().intToRoman(9));
    }
}

File: algorithm-design/leetcode/lc012_integer_to_roman/LC012IntegerToRoman.cs
// https://leetcode.com/problems/integer-to-roman/
using System.Text;

namespace AlgorithmDesign.leetcode.lc012_integer_to_roman
{
    public class Solution
    {
        private readonly string[] _digits = { "M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I" };
        private readonly int[] _weights = { 1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1 };

        public string IntToRoman(int num)
        {
            StringBuilder roman = new StringBuilder();
            for (int i = 0; i < _digits.Length; i++)
            {
                int multiple = num / _weights[i];
                for (int j = 0; j < multiple; j++)
                {
                    roman.Append(_digits[i]);
                }
                num -= multiple * _weights[i];
            }
            return roman.ToString();
        }

    }
}
